java/src/main/java/com/datasqrl/ai/models/bedrock/BedrockChatSession.java

```diff
@@ -0,0 +1,115 @@
+        chatFunctionCall.getArguments()).translate(this::convertExceptionToMessage);
+  }
+
+  @Override
+  public BedrockChatMessage executeFunctionCall(BedrockFunctionCall chatFunctionCall) {
+    try {
+      return new BedrockChatMessage(BedrockChatRole.FUNCTION,
+          backend.executeFunctionCall(chatFunctionCall.getFunctionName(), chatFunctionCall.getArguments(), sessionContext),
+          chatFunctionCall.getFunctionName());
+    } catch (Exception e) {
+      return convertExceptionToMessage(e);
+    }
+  }
+
+  @Override
+  protected BedrockChatMessage convertMessage(GenericChatMessage message) {
+    BedrockChatRole role = BedrockChatRole.valueOf(message.getRole().toUpperCase());
+    //Parse function call?
+    return switch (role) {
+      case SYSTEM -> new BedrockChatMessage(BedrockChatRole.SYSTEM, message.getContent(), message.getName());
+      case USER -> new BedrockChatMessage(BedrockChatRole.USER, message.getContent(), message.getName());
+      case ASSISTANT -> new BedrockChatMessage(BedrockChatRole.ASSISTANT, message.getContent(), message.getName());
+      case FUNCTION -> new BedrockChatMessage(BedrockChatRole.FUNCTION, message.getContent(), message.getName());
+    };
+  }
+
+  @Override
+  protected GenericChatMessage convertMessage(BedrockChatMessage msg) {
+    BedrockFunctionCall fctCall = null;
+    if (msg.getRole() == BedrockChatRole.ASSISTANT) {
+      fctCall = msg.getFunctionCall();
+    }
+    return GenericChatMessage.builder()
+        .role(msg.getRole().getRole())
+        .content(fctCall == null ? msg.getTextContent() : functionCall2String(fctCall))
+        .name(msg.getName())
+        .context(sessionContext)
+        .timestamp(Instant.now().toString())
+        .numTokens(tokenCounter.countTokens(msg))
+        .build();
+  }
+
+  private BedrockChatMessage combineSystemPromptAndFunctions(String systemPrompt) {
+    ObjectMapper objectMapper = new ObjectMapper();
+//    Note: This approach does not take into account the context window for the system prompt
+    String functionText = this.backend.getFunctions().values().stream()
+        .map(RuntimeFunctionDefinition::getChatFunction)
+        .map(f ->
+            objectMapper.createObjectNode()
+                .put("type", "function")
+                .set("function", objectMapper.valueToTree(f))
+        )
+        .map(value -> {
+          try {
+            return objectMapper.writeValueAsString(value);
+          } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+          }
+        })
+        .collect(Collectors.joining("\n"));
+    return new BedrockChatMessage(BedrockChatRole.SYSTEM, systemPrompt + "\n" + functionText + "\n", "");
+  }
+
+  private static String functionCall2String(BedrockFunctionCall fctCall) {
+    return "{"
+        + "\"function\": \"" + fctCall.getFunctionName() + "\", "
+        + "\"parameters\": " + fctCall.getArguments().toString()
+        + "}";
+  }
```

Comment:
@mbroecheler This is the part which I changed to match the function call I described in the system prompt. But I am not sure what the LLM was trained on. Does the function call string description for OpenAIChatSession represent the OpenAPI standard and could Llama3 have been trained on this format?